{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "ChIPseq_vs_CUTNRUN.ipynb",
      "version": "0.3.2",
      "provenance": [],
      "collapsed_sections": [],
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "accelerator": "GPU"
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/kundajelab/chipseqvnewtech/blob/master/H1_POU5F1_NANOG/ChIPseq_vs_CUTNRUN_trainModel.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "WZkK1v0ban-6",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        ""
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "KmDajkfkbJVP",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "%%bash\n",
        "\n",
        "#Download relevant data\n",
        "#For information on data prep, see https://github.com/kundajelab/chipseqvnewtech/blob/master/H1_POU5F1_NANOG/prep_data.sh\n",
        "webdir=http://ec2-13-56-107-121.us-west-1.compute.amazonaws.com/chipseq_vs_newtech/H1_POU5F1_NANOG\n",
        "\n",
        "download_if_not_there () {\n",
        "  local file=$1\n",
        "  [[ -f $file ]] || wget $webdir/$file -O $file\n",
        "}\n",
        "\n",
        "download_if_not_there CUTNRUN.POU5F1.5p.counts.bigWig\n",
        "download_if_not_there CUTNRUN.POU5F1.maxFL120bp.5p.counts.bigWig\n",
        "download_if_not_there CUTNRUN.POU5F1.minFL150bp.5p.counts.bigWig\n",
        "download_if_not_there CUTNRUN.IgG.standard.5p.counts.bigWig\n",
        "download_if_not_there CUTNRUN.IgG.standard.maxFL120bp.5p.counts.bigWig\n",
        "download_if_not_there CUTNRUN.IgG.standard.minFL150bp.5p.counts.bigWig\n",
        "\n",
        "download_if_not_there CUTNRUN.NANOG.5p.counts.bigWig\n",
        "download_if_not_there CUTNRUN.NANOG.maxFL120bp.5p.counts.bigWig\n",
        "download_if_not_there CUTNRUN.NANOG.minFL150bp.5p.counts.bigWig\n",
        "download_if_not_there CUTNRUN.IgG.merged.auto.5p.counts.bigWig\n",
        "download_if_not_there CUTNRUN.IgG.merged.auto.maxFL120bp.5p.counts.bigWig\n",
        "download_if_not_there CUTNRUN.IgG.merged.auto.minFL150bp.5p.counts.bigWig\n",
        "\n",
        "download_if_not_there ChIPseq.POU5F1.merged.5p.counts.plus.bigWig\n",
        "download_if_not_there ChIPseq.POU5F1.merged.5p.counts.minus.bigWig\n",
        "download_if_not_there ChIPseq.NANOG.merged.5p.counts.plus.bigWig\n",
        "download_if_not_there ChIPseq.NANOG.merged.5p.counts.minus.bigWig\n",
        "download_if_not_there ChIPseq.control.bigWig\n",
        "\n",
        "download_if_not_there train_2k_around_summits.bed.gz\n",
        "download_if_not_there valid_2k_around_summits.bed.gz\n",
        "download_if_not_there test_2k_around_summits.bed.gz\n",
        "\n"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "VqpsjI-koJ6P",
        "colab_type": "code",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 391
        },
        "outputId": "e1192fd4-15b3-431e-af64-49284411083a"
      },
      "source": [
        "!ls -ltr"
      ],
      "execution_count": 4,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "total 963764\n",
            "drwxr-xr-x 1 root root      4096 Jun 18 16:14 sample_data\n",
            "-rw-r--r-- 1 root root  72679673 Jul  5 22:31 CUTNRUN.NANOG.5p.counts.bigWig\n",
            "-rw-r--r-- 1 root root  18937954 Jul  5 22:31 CUTNRUN.NANOG.maxFL120bp.5p.counts.bigWig\n",
            "-rw-r--r-- 1 root root  56082417 Jul  5 22:32 CUTNRUN.NANOG.minFL150bp.5p.counts.bigWig\n",
            "-rw-r--r-- 1 root root 139446985 Jul  5 23:32 CUTNRUN.POU5F1.5p.counts.bigWig\n",
            "-rw-r--r-- 1 root root  41550354 Jul  5 23:32 CUTNRUN.POU5F1.maxFL120bp.5p.counts.bigWig\n",
            "-rw-r--r-- 1 root root  78495424 Jul  5 23:32 CUTNRUN.POU5F1.minFL150bp.5p.counts.bigWig\n",
            "-rw-r--r-- 1 root root  43772711 Jul  5 23:56 CUTNRUN.IgG.standard.5p.counts.bigWig\n",
            "-rw-r--r-- 1 root root  23075997 Jul  5 23:57 CUTNRUN.IgG.standard.minFL150bp.5p.counts.bigWig\n",
            "-rw-r--r-- 1 root root  16080996 Jul  5 23:57 CUTNRUN.IgG.standard.maxFL120bp.5p.counts.bigWig\n",
            "-rw-r--r-- 1 root root  51974874 Jul  7 13:16 CUTNRUN.IgG.merged.auto.5p.counts.bigWig\n",
            "-rw-r--r-- 1 root root  35828925 Jul  7 13:16 CUTNRUN.IgG.merged.auto.maxFL120bp.5p.counts.bigWig\n",
            "-rw-r--r-- 1 root root  15424754 Jul  7 13:16 CUTNRUN.IgG.merged.auto.minFL150bp.5p.counts.bigWig\n",
            "-rw-r--r-- 1 root root  77285710 Jul  7 13:17 ChIPseq.POU5F1.merged.5p.counts.plus.bigWig\n",
            "-rw-r--r-- 1 root root  77290787 Jul  7 13:18 ChIPseq.POU5F1.merged.5p.counts.minus.bigWig\n",
            "-rw-r--r-- 1 root root  59203544 Jul  7 13:19 ChIPseq.NANOG.merged.5p.counts.plus.bigWig\n",
            "-rw-r--r-- 1 root root  59221443 Jul  7 13:20 ChIPseq.NANOG.merged.5p.counts.minus.bigWig\n",
            "-rw-r--r-- 1 root root 120086139 Jul  7 13:22 ChIPseq.control.bigWig\n",
            "-rw-r--r-- 1 root root      4924 Jul  7 13:39 valid_2k_around_summits.bed.gz\n",
            "-rw-r--r-- 1 root root    344993 Jul  7 13:39 train_2k_around_summits.bed.gz\n",
            "-rw-r--r-- 1 root root     59041 Jul  7 13:39 test_2k_around_summits.bed.gz\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "QP_gXxlR7YmY",
        "colab_type": "code",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "outputId": "e037b6b5-f8d8-4020-b6e3-f6302ab9ddbc"
      },
      "source": [
        "#install software\n",
        "!pip install pyfaidx\n",
        "!pip install pyBigWig\n",
        "#installing samtools\n",
        "%cd /content\n",
        "![[ -f samtools-1.9.tar.bz2 ]] || wget https://github.com/samtools/samtools/releases/download/1.9/samtools-1.9.tar.bz2\n",
        "!tar -xjf samtools-1.9.tar.bz2\n",
        "%cd samtools-1.9\n",
        "!./configure\n",
        "!make\n",
        "!make install\n",
        "%cd ..\n",
        "\n",
        "#download the hg38 genome, build the index\n",
        "![[ -f hg38.2bit ]] || wget http://hgdownload.cse.ucsc.edu/goldenPath/hg38/bigZips/hg38.2bit -O hg38.2bit  \n",
        "![[ -f twoBitToFa ]] || wget http://hgdownload.soe.ucsc.edu/admin/exe/linux.x86_64/twoBitToFa -O twoBitToFa\n",
        "!chmod a+x twoBitToFa\n",
        "![[ -f hg38.genome.fa ]] || ./twoBitToFa hg38.2bit hg38.genome.fa\n",
        "![[ -f hg38.genome.fa.fai ]] || samtools faidx hg38.genome.fa\n",
        "#Download the ENCODE ebv-containing chromsizes file\n",
        "![[ -f hg38.chrom.sizes ]] || wget https://raw.githubusercontent.com/ENCODE-DCC/encValData/master/GRCh38/GRCh38_EBV.chrom.sizes -O hg38.chrom.sizes"
      ],
      "execution_count": 5,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Collecting pyfaidx\n",
            "  Downloading https://files.pythonhosted.org/packages/75/a5/7e2569527b3849ea28d79b4f70d7cf46a47d36459bc59e0efa4e10e8c8b2/pyfaidx-0.5.5.2.tar.gz\n",
            "Requirement already satisfied: six in /usr/local/lib/python3.6/dist-packages (from pyfaidx) (1.12.0)\n",
            "Requirement already satisfied: setuptools>=0.7 in /usr/local/lib/python3.6/dist-packages (from pyfaidx) (41.0.1)\n",
            "Building wheels for collected packages: pyfaidx\n",
            "  Building wheel for pyfaidx (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Stored in directory: /root/.cache/pip/wheels/54/a2/b4/e242e58d23b2808e191b214067880faa46cd2341f363886e0b\n",
            "Successfully built pyfaidx\n",
            "Installing collected packages: pyfaidx\n",
            "Successfully installed pyfaidx-0.5.5.2\n",
            "Collecting pyBigWig\n",
            "\u001b[?25l  Downloading https://files.pythonhosted.org/packages/67/c8/741cca63e6a8688ef22a605bb06cc849fc57df419027013f2d1181f184f8/pyBigWig-0.3.16.tar.gz (65kB)\n",
            "\u001b[K     |████████████████████████████████| 71kB 4.7MB/s \n",
            "\u001b[?25hBuilding wheels for collected packages: pyBigWig\n",
            "  Building wheel for pyBigWig (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Stored in directory: /root/.cache/pip/wheels/d7/b9/a7/4ced1fb3a4e8816bbdc7bd3fee1d378d76e57e55ead6113976\n",
            "Successfully built pyBigWig\n",
            "Installing collected packages: pyBigWig\n",
            "Successfully installed pyBigWig-0.3.16\n",
            "/content\n",
            "--2019-07-07 13:42:24--  https://github.com/samtools/samtools/releases/download/1.9/samtools-1.9.tar.bz2\n",
            "Resolving github.com (github.com)... 140.82.114.3\n",
            "Connecting to github.com (github.com)|140.82.114.3|:443... connected.\n",
            "HTTP request sent, awaiting response... 302 Found\n",
            "Location: https://github-production-release-asset-2e65be.s3.amazonaws.com/3666841/fe586164-8a73-11e8-84ad-bb90bbd3b7c0?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20190707%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20190707T134225Z&X-Amz-Expires=300&X-Amz-Signature=98bb5d777eb4393c80e43c0214ba9035aa3061ba16c9c468e33a9b452b92fe12&X-Amz-SignedHeaders=host&actor_id=0&response-content-disposition=attachment%3B%20filename%3Dsamtools-1.9.tar.bz2&response-content-type=application%2Foctet-stream [following]\n",
            "--2019-07-07 13:42:25--  https://github-production-release-asset-2e65be.s3.amazonaws.com/3666841/fe586164-8a73-11e8-84ad-bb90bbd3b7c0?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20190707%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20190707T134225Z&X-Amz-Expires=300&X-Amz-Signature=98bb5d777eb4393c80e43c0214ba9035aa3061ba16c9c468e33a9b452b92fe12&X-Amz-SignedHeaders=host&actor_id=0&response-content-disposition=attachment%3B%20filename%3Dsamtools-1.9.tar.bz2&response-content-type=application%2Foctet-stream\n",
            "Resolving github-production-release-asset-2e65be.s3.amazonaws.com (github-production-release-asset-2e65be.s3.amazonaws.com)... 52.216.145.115\n",
            "Connecting to github-production-release-asset-2e65be.s3.amazonaws.com (github-production-release-asset-2e65be.s3.amazonaws.com)|52.216.145.115|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 4440405 (4.2M) [application/octet-stream]\n",
            "Saving to: ‘samtools-1.9.tar.bz2’\n",
            "\n",
            "samtools-1.9.tar.bz 100%[===================>]   4.23M  9.53MB/s    in 0.4s    \n",
            "\n",
            "2019-07-07 13:42:26 (9.53 MB/s) - ‘samtools-1.9.tar.bz2’ saved [4440405/4440405]\n",
            "\n",
            "/content/samtools-1.9\n",
            "checking for gcc... gcc\n",
            "checking whether the C compiler works... yes\n",
            "checking for C compiler default output file name... a.out\n",
            "checking for suffix of executables... \n",
            "checking whether we are cross compiling... no\n",
            "checking for suffix of object files... o\n",
            "checking whether we are using the GNU C compiler... yes\n",
            "checking whether gcc accepts -g... yes\n",
            "checking for gcc option to accept ISO C89... none needed\n",
            "checking for special C compiler options needed for large files... no\n",
            "checking for _FILE_OFFSET_BITS value needed for large files... no\n",
            "checking location of HTSlib source tree... htslib-1.9\n",
            "checking for NcursesW wide-character library... no\n",
            "checking for Ncurses library... yes\n",
            "checking for working ncurses/curses.h... no\n",
            "checking for working ncurses.h... yes\n",
            "checking for zlib.h... yes\n",
            "checking for inflate in -lz... yes\n",
            "checking for library containing regcomp... none required\n",
            "configure: creating ./config.status\n",
            "config.status: creating config.mk\n",
            "config.status: creating config.h\n",
            "=== configuring in htslib-1.9 (/content/samtools-1.9/htslib-1.9)\n",
            "configure: running /bin/bash ./configure --disable-option-checking '--prefix=/usr/local'  --cache-file=/dev/null --srcdir=.\n",
            "checking for gcc... gcc\n",
            "checking whether the C compiler works... yes\n",
            "checking for C compiler default output file name... a.out\n",
            "checking for suffix of executables... \n",
            "checking whether we are cross compiling... no\n",
            "checking for suffix of object files... o\n",
            "checking whether we are using the GNU C compiler... yes\n",
            "checking whether gcc accepts -g... yes\n",
            "checking for gcc option to accept ISO C89... none needed\n",
            "checking for ranlib... ranlib\n",
            "checking for grep that handles long lines and -e... /bin/grep\n",
            "checking for C compiler warning flags... -Wall\n",
            "checking for special C compiler options needed for large files... no\n",
            "checking for _FILE_OFFSET_BITS value needed for large files... no\n",
            "checking for _LARGEFILE_SOURCE value needed for large files... no\n",
            "checking shared library type for unknown-Linux... plain .so\n",
            "checking how to run the C preprocessor... gcc -E\n",
            "checking for egrep... /bin/grep -E\n",
            "checking for ANSI C header files... yes\n",
            "checking for sys/types.h... yes\n",
            "checking for sys/stat.h... yes\n",
            "checking for stdlib.h... yes\n",
            "checking for string.h... yes\n",
            "checking for memory.h... yes\n",
            "checking for strings.h... yes\n",
            "checking for inttypes.h... yes\n",
            "checking for stdint.h... yes\n",
            "checking for unistd.h... yes\n",
            "checking for stdlib.h... (cached) yes\n",
            "checking for unistd.h... (cached) yes\n",
            "checking for sys/param.h... yes\n",
            "checking for getpagesize... yes\n",
            "checking for working mmap... yes\n",
            "checking for gmtime_r... yes\n",
            "checking for fsync... yes\n",
            "checking for drand48... yes\n",
            "checking whether fdatasync is declared... yes\n",
            "checking for fdatasync... yes\n",
            "checking for library containing log... -lm\n",
            "checking for zlib.h... yes\n",
            "checking for inflate in -lz... yes\n",
            "checking for library containing recv... none required\n",
            "checking for bzlib.h... yes\n",
            "checking for BZ2_bzBuffToBuffCompress in -lbz2... yes\n",
            "checking for lzma.h... yes\n",
            "checking for lzma_easy_buffer_encode in -llzma... yes\n",
            "checking for libdeflate.h... no\n",
            "checking for libdeflate_deflate_compress in -ldeflate... no\n",
            "checking for curl_easy_pause in -lcurl... yes\n",
            "checking for CCHmac... no\n",
            "checking for library containing HMAC... -lcrypto\n",
            "checking whether PTHREAD_MUTEX_RECURSIVE is declared... yes\n",
            "configure: creating ./config.status\n",
            "config.status: creating config.mk\n",
            "config.status: creating htslib.pc.tmp\n",
            "config.status: creating config.h\n",
            "config.mk:45: htslib-1.9/htslib_static.mk: No such file or directory\n",
            "cd htslib-1.9 && make htslib_static.mk\n",
            "make[1]: Entering directory '/content/samtools-1.9/htslib-1.9'\n",
            "sed -n '/^static_libs=/s/[^=]*=/HTSLIB_static_LIBS = /p;/^static_ldflags=/s/[^=]*=/HTSLIB_static_LDFLAGS = /p' htslib.pc.tmp > htslib_static.mk\n",
            "make[1]: Leaving directory '/content/samtools-1.9/htslib-1.9'\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o bam_index.o bam_index.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o bam_plcmd.o bam_plcmd.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o sam_view.o sam_view.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o bam_cat.o bam_cat.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o bam_md.o bam_md.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o bam_reheader.o bam_reheader.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o bam_sort.o bam_sort.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o bedidx.o bedidx.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o bam_rmdup.o bam_rmdup.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o bam_rmdupse.o bam_rmdupse.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o bam_mate.o bam_mate.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o bam_stat.o bam_stat.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o bam_color.o bam_color.c\n",
            "echo '#define SAMTOOLS_VERSION \"1.9\"' > version.h\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o bamtk.o bamtk.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o bam2bcf.o bam2bcf.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o bam2bcf_indel.o bam2bcf_indel.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o sample.o sample.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o cut_target.o cut_target.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o phase.o phase.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o bam2depth.o bam2depth.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o padding.o padding.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o bedcov.o bedcov.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o bamshuf.o bamshuf.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o faidx.o faidx.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o dict.o dict.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o stats.o stats.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o stats_isize.o stats_isize.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o bam_flags.o bam_flags.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o bam_split.o bam_split.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o bam_tview.o bam_tview.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o bam_tview_curses.o bam_tview_curses.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o bam_tview_html.o bam_tview_html.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o bam_lpileup.o bam_lpileup.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o bam_quickcheck.o bam_quickcheck.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o bam_addrprg.o bam_addrprg.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o bam_markdup.o bam_markdup.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o tmp_file.o tmp_file.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o lz4/lz4.o lz4/lz4.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o bam_aux.o bam_aux.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o bam.o bam.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o bam_import.o bam_import.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o sam.o sam.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o sam_header.o sam_header.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o bam_plbuf.o bam_plbuf.c\n",
            "ar -csru libbam.a bam_aux.o bam.o bam_import.o sam.o sam_header.o bam_plbuf.o\n",
            "ar: `u' modifier ignored since `D' is the default (see `U')\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o sam_opts.o sam_opts.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o sam_utils.o sam_utils.c\n",
            "ar -rcs libst.a sam_opts.o sam_utils.o\n",
            "cd htslib-1.9 && make lib-static\n",
            "make[1]: Entering directory '/content/samtools-1.9/htslib-1.9'\n",
            "gcc -Wall -g -O2 -I.  -c -o kfunc.o kfunc.c\n",
            "gcc -Wall -g -O2 -I.  -c -o knetfile.o knetfile.c\n",
            "gcc -Wall -g -O2 -I.  -c -o kstring.o kstring.c\n",
            "gcc -Wall -g -O2 -I.  -c -o bcf_sr_sort.o bcf_sr_sort.c\n",
            "gcc -Wall -g -O2 -I.  -c -o bgzf.o bgzf.c\n",
            "gcc -Wall -g -O2 -I.  -c -o errmod.o errmod.c\n",
            "gcc -Wall -g -O2 -I.  -c -o faidx.o faidx.c\n",
            "gcc -Wall -g -O2 -I.  -c -o hfile.o hfile.c\n",
            "gcc -Wall -g -O2 -I.  -c -o hfile_net.o hfile_net.c\n",
            "echo '#define HTS_VERSION \"1.9\"' > version.h\n",
            "gcc -Wall -g -O2 -I.  -c -o hts.o hts.c\n",
            "gcc -Wall -g -O2 -I.  -c -o hts_os.o hts_os.c\n",
            "gcc -Wall -g -O2 -I.  -c -o md5.o md5.c\n",
            "gcc -Wall -g -O2 -I.  -c -o multipart.o multipart.c\n",
            "gcc -Wall -g -O2 -I.  -c -o probaln.o probaln.c\n",
            "gcc -Wall -g -O2 -I.  -c -o realn.o realn.c\n",
            "gcc -Wall -g -O2 -I.  -c -o regidx.o regidx.c\n",
            "gcc -Wall -g -O2 -I.  -c -o sam.o sam.c\n",
            "gcc -Wall -g -O2 -I.  -c -o synced_bcf_reader.o synced_bcf_reader.c\n",
            "gcc -Wall -g -O2 -I.  -c -o vcf_sweep.o vcf_sweep.c\n",
            "gcc -Wall -g -O2 -I.  -c -o tbx.o tbx.c\n",
            "gcc -Wall -g -O2 -I.  -c -o textutils.o textutils.c\n",
            "gcc -Wall -g -O2 -I.  -c -o thread_pool.o thread_pool.c\n",
            "gcc -Wall -g -O2 -I.  -c -o vcf.o vcf.c\n",
            "gcc -Wall -g -O2 -I.  -c -o vcfutils.o vcfutils.c\n",
            "gcc -Wall -g -O2 -I.  -c -o cram/cram_codecs.o cram/cram_codecs.c\n",
            "gcc -Wall -g -O2 -I.  -c -o cram/cram_decode.o cram/cram_decode.c\n",
            "gcc -Wall -g -O2 -I.  -c -o cram/cram_encode.o cram/cram_encode.c\n",
            "gcc -Wall -g -O2 -I.  -c -o cram/cram_external.o cram/cram_external.c\n",
            "gcc -Wall -g -O2 -I.  -c -o cram/cram_index.o cram/cram_index.c\n",
            "gcc -Wall -g -O2 -I.  -c -o cram/cram_io.o cram/cram_io.c\n",
            "gcc -Wall -g -O2 -I.  -c -o cram/cram_samtools.o cram/cram_samtools.c\n",
            "gcc -Wall -g -O2 -I.  -c -o cram/cram_stats.o cram/cram_stats.c\n",
            "gcc -Wall -g -O2 -I.  -c -o cram/files.o cram/files.c\n",
            "gcc -Wall -g -O2 -I.  -c -o cram/mFILE.o cram/mFILE.c\n",
            "gcc -Wall -g -O2 -I.  -c -o cram/open_trace_file.o cram/open_trace_file.c\n",
            "gcc -Wall -g -O2 -I.  -c -o cram/pooled_alloc.o cram/pooled_alloc.c\n",
            "gcc -Wall -g -O2 -I.  -c -o cram/rANS_static.o cram/rANS_static.c\n",
            "gcc -Wall -g -O2 -I.  -c -o cram/sam_header.o cram/sam_header.c\n",
            "gcc -Wall -g -O2 -I.  -c -o cram/string_alloc.o cram/string_alloc.c\n",
            "gcc -Wall -g -O2 -I.  -c -o hfile_libcurl.o hfile_libcurl.c\n",
            "gcc -Wall -g -O2 -I.  -c -o hfile_gcs.o hfile_gcs.c\n",
            "gcc -Wall -g -O2 -I.  -c -o hfile_s3.o hfile_s3.c\n",
            "ar -rc libhts.a kfunc.o knetfile.o kstring.o bcf_sr_sort.o bgzf.o errmod.o faidx.o hfile.o hfile_net.o hts.o hts_os.o md5.o multipart.o probaln.o realn.o regidx.o sam.o synced_bcf_reader.o vcf_sweep.o tbx.o textutils.o thread_pool.o vcf.o vcfutils.o cram/cram_codecs.o cram/cram_decode.o cram/cram_encode.o cram/cram_external.o cram/cram_index.o cram/cram_io.o cram/cram_samtools.o cram/cram_stats.o cram/files.o cram/mFILE.o cram/open_trace_file.o cram/pooled_alloc.o cram/rANS_static.o cram/sam_header.o cram/string_alloc.o  hfile_libcurl.o hfile_gcs.o hfile_s3.o\n",
            "ranlib libhts.a\n",
            "make[1]: Leaving directory '/content/samtools-1.9/htslib-1.9'\n",
            "gcc  -L./lz4  -o samtools bam_index.o bam_plcmd.o sam_view.o bam_cat.o bam_md.o bam_reheader.o bam_sort.o bedidx.o bam_rmdup.o bam_rmdupse.o bam_mate.o bam_stat.o bam_color.o bamtk.o bam2bcf.o bam2bcf_indel.o sample.o cut_target.o phase.o bam2depth.o padding.o bedcov.o bamshuf.o faidx.o dict.o stats.o stats_isize.o bam_flags.o bam_split.o bam_tview.o bam_tview_curses.o bam_tview_html.o bam_lpileup.o bam_quickcheck.o bam_addrprg.o bam_markdup.o tmp_file.o ./lz4/lz4.o libbam.a libst.a htslib-1.9/libhts.a -lpthread -lz -lm -lbz2 -llzma -lcurl -lcrypto -lncurses -lm -lz  -lpthread\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o misc/ace2sam.o misc/ace2sam.c\n",
            "gcc  -o misc/ace2sam misc/ace2sam.o -lz \n",
            "gcc -g -O2 -DMAQ_LONGREADS -I. -Ihtslib-1.9 -I./lz4  -c -o misc/maq2sam-long.o misc/maq2sam.c\n",
            "gcc  -o misc/maq2sam-long misc/maq2sam-long.o -lz \n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o misc/maq2sam-short.o misc/maq2sam.c\n",
            "gcc  -o misc/maq2sam-short misc/maq2sam-short.o -lz \n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o misc/md5fa.o misc/md5fa.c\n",
            "gcc  -L./lz4  -o misc/md5fa misc/md5fa.o htslib-1.9/libhts.a -lpthread -lz -lm -lbz2 -llzma -lcurl -lcrypto -lz \n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o misc/md5sum-lite.o misc/md5sum-lite.c\n",
            "gcc  -L./lz4  -o misc/md5sum-lite misc/md5sum-lite.o htslib-1.9/libhts.a -lpthread -lz -lm -lbz2 -llzma -lcurl -lcrypto -lz \n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o misc/wgsim.o misc/wgsim.c\n",
            "gcc  -L./lz4  -o misc/wgsim misc/wgsim.o -lm htslib-1.9/libhts.a -lpthread -lz -lm -lbz2 -llzma -lcurl -lcrypto -lz \n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o test/merge/test_bam_translate.o test/merge/test_bam_translate.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o test/test.o test/test.c\n",
            "gcc  -L./lz4  -o test/merge/test_bam_translate test/merge/test_bam_translate.o test/test.o libst.a htslib-1.9/libhts.a -lpthread -lz -lm -lbz2 -llzma -lcurl -lcrypto -lz  -lpthread\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o test/merge/test_rtrans_build.o test/merge/test_rtrans_build.c\n",
            "gcc  -L./lz4  -o test/merge/test_rtrans_build test/merge/test_rtrans_build.o libst.a htslib-1.9/libhts.a -lpthread -lz -lm -lbz2 -llzma -lcurl -lcrypto -lz  -lpthread\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o test/merge/test_trans_tbl_init.o test/merge/test_trans_tbl_init.c\n",
            "gcc  -L./lz4  -o test/merge/test_trans_tbl_init test/merge/test_trans_tbl_init.o libst.a htslib-1.9/libhts.a -lpthread -lz -lm -lbz2 -llzma -lcurl -lcrypto -lz  -lpthread\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o test/split/test_count_rg.o test/split/test_count_rg.c\n",
            "gcc  -L./lz4  -o test/split/test_count_rg test/split/test_count_rg.o test/test.o libst.a htslib-1.9/libhts.a -lpthread -lz -lm -lbz2 -llzma -lcurl -lcrypto -lz  -lpthread\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o test/split/test_expand_format_string.o test/split/test_expand_format_string.c\n",
            "gcc  -L./lz4  -o test/split/test_expand_format_string test/split/test_expand_format_string.o test/test.o libst.a htslib-1.9/libhts.a -lpthread -lz -lm -lbz2 -llzma -lcurl -lcrypto -lz  -lpthread\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o test/split/test_filter_header_rg.o test/split/test_filter_header_rg.c\n",
            "gcc  -L./lz4  -o test/split/test_filter_header_rg test/split/test_filter_header_rg.o test/test.o libst.a htslib-1.9/libhts.a -lpthread -lz -lm -lbz2 -llzma -lcurl -lcrypto -lz  -lpthread\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o test/split/test_parse_args.o test/split/test_parse_args.c\n",
            "gcc  -L./lz4  -o test/split/test_parse_args test/split/test_parse_args.o test/test.o libst.a htslib-1.9/libhts.a -lpthread -lz -lm -lbz2 -llzma -lcurl -lcrypto -lz  -lpthread\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o test/vcf-miniview.o test/vcf-miniview.c\n",
            "gcc  -L./lz4  -o test/vcf-miniview test/vcf-miniview.o htslib-1.9/libhts.a -lpthread -lz -lm -lbz2 -llzma -lcurl -lcrypto -lz  -lpthread\n",
            "mkdir -p -m 755 /usr/local/bin /usr/local/bin /usr/local/share/man/man1\n",
            "install -p samtools /usr/local/bin\n",
            "install -p misc/ace2sam misc/maq2sam-long misc/maq2sam-short misc/md5fa misc/md5sum-lite misc/wgsim /usr/local/bin\n",
            "install -p misc/blast2sam.pl misc/bowtie2sam.pl misc/export2sam.pl misc/interpolate_sam.pl misc/novo2sam.pl misc/plot-bamstats misc/psl2sam.pl misc/sam2vcf.pl misc/samtools.pl misc/seq_cache_populate.pl misc/soap2sam.pl misc/varfilter.py misc/wgsim_eval.pl misc/zoom2sam.pl /usr/local/bin\n",
            "install -p -m 644 samtools.1 misc/wgsim.1 /usr/local/share/man/man1\n",
            "/content\n",
            "--2019-07-07 13:43:23--  http://hgdownload.cse.ucsc.edu/goldenPath/hg38/bigZips/hg38.2bit\n",
            "Resolving hgdownload.cse.ucsc.edu (hgdownload.cse.ucsc.edu)... 128.114.119.163\n",
            "Connecting to hgdownload.cse.ucsc.edu (hgdownload.cse.ucsc.edu)|128.114.119.163|:80... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 835393456 (797M) [text/plain]\n",
            "Saving to: ‘hg38.2bit’\n",
            "\n",
            "hg38.2bit           100%[===================>] 796.69M   125MB/s    in 8.5s    \n",
            "\n",
            "2019-07-07 13:43:31 (93.7 MB/s) - ‘hg38.2bit’ saved [835393456/835393456]\n",
            "\n",
            "--2019-07-07 13:43:32--  http://hgdownload.soe.ucsc.edu/admin/exe/linux.x86_64/twoBitToFa\n",
            "Resolving hgdownload.soe.ucsc.edu (hgdownload.soe.ucsc.edu)... 128.114.119.163\n",
            "Connecting to hgdownload.soe.ucsc.edu (hgdownload.soe.ucsc.edu)|128.114.119.163|:80... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 5884080 (5.6M) [text/plain]\n",
            "Saving to: ‘twoBitToFa’\n",
            "\n",
            "twoBitToFa          100%[===================>]   5.61M  5.31MB/s    in 1.1s    \n",
            "\n",
            "2019-07-07 13:43:33 (5.31 MB/s) - ‘twoBitToFa’ saved [5884080/5884080]\n",
            "\n",
            "--2019-07-07 13:44:11--  https://raw.githubusercontent.com/ENCODE-DCC/encValData/master/GRCh38/GRCh38_EBV.chrom.sizes\n",
            "Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 151.101.0.133, 151.101.64.133, 151.101.128.133, ...\n",
            "Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|151.101.0.133|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 11686 (11K) [text/plain]\n",
            "Saving to: ‘hg38.chrom.sizes’\n",
            "\n",
            "hg38.chrom.sizes    100%[===================>]  11.41K  --.-KB/s    in 0s      \n",
            "\n",
            "2019-07-07 13:44:11 (134 MB/s) - ‘hg38.chrom.sizes’ saved [11686/11686]\n",
            "\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "nc8NEIM3OlKp",
        "colab_type": "code",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "outputId": "f24d66bd-2f31-45db-c2fb-7faa6e1e0363"
      },
      "source": [
        "![[ -e seqdataloader ]] && rm -rf seqdataloader\n",
        "!git clone https://github.com/kundajelab/seqdataloader.git\n",
        "!pip uninstall seqdataloader\n",
        "!pip install seqdataloader/"
      ],
      "execution_count": 1,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Cloning into 'seqdataloader'...\n",
            "remote: Enumerating objects: 67, done.\u001b[K\n",
            "remote: Counting objects:   1% (1/67)   \u001b[K\rremote: Counting objects:   2% (2/67)   \u001b[K\rremote: Counting objects:   4% (3/67)   \u001b[K\rremote: Counting objects:   5% (4/67)   \u001b[K\rremote: Counting objects:   7% (5/67)   \u001b[K\rremote: Counting objects:   8% (6/67)   \u001b[K\rremote: Counting objects:  10% (7/67)   \u001b[K\rremote: Counting objects:  11% (8/67)   \u001b[K\rremote: Counting objects:  13% (9/67)   \u001b[K\rremote: Counting objects:  14% (10/67)   \u001b[K\rremote: Counting objects:  16% (11/67)   \u001b[K\rremote: Counting objects:  17% (12/67)   \u001b[K\rremote: Counting objects:  19% (13/67)   \u001b[K\rremote: Counting objects:  20% (14/67)   \u001b[K\rremote: Counting objects:  22% (15/67)   \u001b[K\rremote: Counting objects:  23% (16/67)   \u001b[K\rremote: Counting objects:  25% (17/67)   \u001b[K\rremote: Counting objects:  26% (18/67)   \u001b[K\rremote: Counting objects:  28% (19/67)   \u001b[K\rremote: Counting objects:  29% (20/67)   \u001b[K\rremote: Counting objects:  31% (21/67)   \u001b[K\rremote: Counting objects:  32% (22/67)   \u001b[K\rremote: Counting objects:  34% (23/67)   \u001b[K\rremote: Counting objects:  35% (24/67)   \u001b[K\rremote: Counting objects:  37% (25/67)   \u001b[K\rremote: Counting objects:  38% (26/67)   \u001b[K\rremote: Counting objects:  40% (27/67)   \u001b[K\rremote: Counting objects:  41% (28/67)   \u001b[K\rremote: Counting objects:  43% (29/67)   \u001b[K\rremote: Counting objects:  44% (30/67)   \u001b[K\rremote: Counting objects:  46% (31/67)   \u001b[K\rremote: Counting objects:  47% (32/67)   \u001b[K\rremote: Counting objects:  49% (33/67)   \u001b[K\rremote: Counting objects:  50% (34/67)   \u001b[K\rremote: Counting objects:  52% (35/67)   \u001b[K\rremote: Counting objects:  53% (36/67)   \u001b[K\rremote: Counting objects:  55% (37/67)   \u001b[K\rremote: Counting objects:  56% (38/67)   \u001b[K\rremote: Counting objects:  58% (39/67)   \u001b[K\rremote: Counting objects:  59% (40/67)   \u001b[K\rremote: Counting objects:  61% (41/67)   \u001b[K\rremote: Counting objects:  62% (42/67)   \u001b[K\rremote: Counting objects:  64% (43/67)   \u001b[K\rremote: Counting objects:  65% (44/67)   \u001b[K\rremote: Counting objects:  67% (45/67)   \u001b[K\rremote: Counting objects:  68% (46/67)   \u001b[K\rremote: Counting objects:  70% (47/67)   \u001b[K\rremote: Counting objects:  71% (48/67)   \u001b[K\rremote: Counting objects:  73% (49/67)   \u001b[K\rremote: Counting objects:  74% (50/67)   \u001b[K\rremote: Counting objects:  76% (51/67)   \u001b[K\rremote: Counting objects:  77% (52/67)   \u001b[K\rremote: Counting objects:  79% (53/67)   \u001b[K\rremote: Counting objects:  80% (54/67)   \u001b[K\rremote: Counting objects:  82% (55/67)   \u001b[K\rremote: Counting objects:  83% (56/67)   \u001b[K\rremote: Counting objects:  85% (57/67)   \u001b[K\rremote: Counting objects:  86% (58/67)   \u001b[K\rremote: Counting objects:  88% (59/67)   \u001b[K\rremote: Counting objects:  89% (60/67)   \u001b[K\rremote: Counting objects:  91% (61/67)   \u001b[K\rremote: Counting objects:  92% (62/67)   \u001b[K\rremote: Counting objects:  94% (63/67)   \u001b[K\rremote: Counting objects:  95% (64/67)   \u001b[K\rremote: Counting objects:  97% (65/67)   \u001b[K\rremote: Counting objects:  98% (66/67)   \u001b[K\rremote: Counting objects: 100% (67/67)   \u001b[K\rremote: Counting objects: 100% (67/67), done.\u001b[K\n",
            "remote: Compressing objects:   2% (1/46)   \u001b[K\rremote: Compressing objects:   4% (2/46)   \u001b[K\rremote: Compressing objects:   6% (3/46)   \u001b[K\rremote: Compressing objects:   8% (4/46)   \u001b[K\rremote: Compressing objects:  10% (5/46)   \u001b[K\rremote: Compressing objects:  13% (6/46)   \u001b[K\rremote: Compressing objects:  15% (7/46)   \u001b[K\rremote: Compressing objects:  17% (8/46)   \u001b[K\rremote: Compressing objects:  19% (9/46)   \u001b[K\rremote: Compressing objects:  21% (10/46)   \u001b[K\rremote: Compressing objects:  23% (11/46)   \u001b[K\rremote: Compressing objects:  26% (12/46)   \u001b[K\rremote: Compressing objects:  28% (13/46)   \u001b[K\rremote: Compressing objects:  30% (14/46)   \u001b[K\rremote: Compressing objects:  32% (15/46)   \u001b[K\rremote: Compressing objects:  34% (16/46)   \u001b[K\rremote: Compressing objects:  36% (17/46)   \u001b[K\rremote: Compressing objects:  39% (18/46)   \u001b[K\rremote: Compressing objects:  41% (19/46)   \u001b[K\rremote: Compressing objects:  43% (20/46)   \u001b[K\rremote: Compressing objects:  45% (21/46)   \u001b[K\rremote: Compressing objects:  47% (22/46)   \u001b[K\rremote: Compressing objects:  50% (23/46)   \u001b[K\rremote: Compressing objects:  52% (24/46)   \u001b[K\rremote: Compressing objects:  54% (25/46)   \u001b[K\rremote: Compressing objects:  56% (26/46)   \u001b[K\rremote: Compressing objects:  58% (27/46)   \u001b[K\rremote: Compressing objects:  60% (28/46)   \u001b[K\rremote: Compressing objects:  63% (29/46)   \u001b[K\rremote: Compressing objects:  65% (30/46)   \u001b[K\rremote: Compressing objects:  67% (31/46)   \u001b[K\rremote: Compressing objects:  69% (32/46)   \u001b[K\rremote: Compressing objects:  71% (33/46)   \u001b[K\rremote: Compressing objects:  73% (34/46)   \u001b[K\rremote: Compressing objects:  76% (35/46)   \u001b[K\rremote: Compressing objects:  78% (36/46)   \u001b[K\rremote: Compressing objects:  80% (37/46)   \u001b[K\rremote: Compressing objects:  82% (38/46)   \u001b[K\rremote: Compressing objects:  84% (39/46)   \u001b[K\rremote: Compressing objects:  86% (40/46)   \u001b[K\rremote: Compressing objects:  89% (41/46)   \u001b[K\rremote: Compressing objects:  91% (42/46)   \u001b[K\rremote: Compressing objects:  93% (43/46)   \u001b[K\rremote: Compressing objects:  95% (44/46)   \u001b[K\rremote: Compressing objects:  97% (45/46)   \u001b[K\rremote: Compressing objects: 100% (46/46)   \u001b[K\rremote: Compressing objects: 100% (46/46), done.\u001b[K\n",
            "Receiving objects:   0% (1/575)   \rReceiving objects:   1% (6/575)   \rReceiving objects:   2% (12/575)   \rReceiving objects:   3% (18/575)   \rReceiving objects:   4% (23/575)   \rReceiving objects:   5% (29/575)   \rReceiving objects:   6% (35/575)   \rReceiving objects:   7% (41/575)   \rReceiving objects:   8% (46/575)   \rReceiving objects:   9% (52/575)   \rReceiving objects:  10% (58/575)   \rReceiving objects:  11% (64/575)   \rReceiving objects:  12% (69/575)   \rReceiving objects:  13% (75/575)   \rReceiving objects:  14% (81/575)   \rReceiving objects:  15% (87/575)   \rReceiving objects:  16% (92/575)   \rReceiving objects:  17% (98/575)   \rReceiving objects:  18% (104/575)   \rReceiving objects:  19% (110/575)   \rReceiving objects:  20% (115/575)   \rReceiving objects:  21% (121/575)   \rReceiving objects:  22% (127/575)   \rReceiving objects:  23% (133/575)   \rReceiving objects:  24% (138/575)   \rReceiving objects:  25% (144/575)   \rReceiving objects:  26% (150/575)   \rReceiving objects:  27% (156/575)   \rReceiving objects:  28% (161/575)   \rReceiving objects:  29% (167/575)   \rReceiving objects:  30% (173/575)   \rReceiving objects:  31% (179/575)   \rReceiving objects:  32% (184/575)   \rReceiving objects:  33% (190/575)   \rReceiving objects:  34% (196/575)   \rReceiving objects:  35% (202/575)   \rReceiving objects:  36% (207/575)   \rReceiving objects:  37% (213/575)   \rReceiving objects:  38% (219/575)   \rReceiving objects:  39% (225/575)   \rReceiving objects:  40% (230/575)   \rReceiving objects:  41% (236/575)   \rReceiving objects:  42% (242/575)   \rReceiving objects:  43% (248/575)   \rReceiving objects:  44% (253/575)   \rReceiving objects:  45% (259/575)   \rReceiving objects:  46% (265/575)   \rReceiving objects:  47% (271/575)   \rReceiving objects:  48% (276/575)   \rReceiving objects:  49% (282/575)   \rReceiving objects:  50% (288/575)   \rReceiving objects:  51% (294/575)   \rReceiving objects:  52% (299/575)   \rReceiving objects:  53% (305/575)   \rReceiving objects:  54% (311/575)   \rReceiving objects:  55% (317/575)   \rReceiving objects:  56% (322/575)   \rReceiving objects:  57% (328/575)   \rReceiving objects:  58% (334/575)   \rReceiving objects:  59% (340/575)   \rReceiving objects:  60% (345/575)   \rReceiving objects:  61% (351/575)   \rReceiving objects:  62% (357/575)   \rReceiving objects:  63% (363/575)   \rReceiving objects:  64% (368/575)   \rReceiving objects:  65% (374/575)   \rReceiving objects:  66% (380/575)   \rReceiving objects:  67% (386/575)   \rReceiving objects:  68% (391/575)   \rReceiving objects:  69% (397/575)   \rReceiving objects:  70% (403/575)   \rReceiving objects:  71% (409/575)   \rReceiving objects:  72% (414/575)   \rReceiving objects:  73% (420/575)   \rReceiving objects:  74% (426/575)   \rReceiving objects:  75% (432/575)   \rremote: Total 575 (delta 41), reused 44 (delta 21), pack-reused 508\u001b[K\n",
            "Receiving objects:  76% (437/575)   \rReceiving objects:  77% (443/575)   \rReceiving objects:  78% (449/575)   \rReceiving objects:  79% (455/575)   \rReceiving objects:  80% (460/575)   \rReceiving objects:  81% (466/575)   \rReceiving objects:  82% (472/575)   \rReceiving objects:  83% (478/575)   \rReceiving objects:  84% (483/575)   \rReceiving objects:  85% (489/575)   \rReceiving objects:  86% (495/575)   \rReceiving objects:  87% (501/575)   \rReceiving objects:  88% (506/575)   \rReceiving objects:  89% (512/575)   \rReceiving objects:  90% (518/575)   \rReceiving objects:  91% (524/575)   \rReceiving objects:  92% (529/575)   \rReceiving objects:  93% (535/575)   \rReceiving objects:  94% (541/575)   \rReceiving objects:  95% (547/575)   \rReceiving objects:  96% (552/575)   \rReceiving objects:  97% (558/575)   \rReceiving objects:  98% (564/575)   \rReceiving objects:  99% (570/575)   \rReceiving objects: 100% (575/575)   \rReceiving objects: 100% (575/575), 3.80 MiB | 49.83 MiB/s, done.\n",
            "Resolving deltas:   0% (0/367)   \rResolving deltas:   5% (22/367)   \rResolving deltas:   6% (25/367)   \rResolving deltas:  19% (70/367)   \rResolving deltas:  21% (80/367)   \rResolving deltas:  22% (81/367)   \rResolving deltas:  23% (86/367)   \rResolving deltas:  30% (112/367)   \rResolving deltas:  32% (119/367)   \rResolving deltas:  33% (124/367)   \rResolving deltas:  35% (129/367)   \rResolving deltas:  38% (140/367)   \rResolving deltas:  40% (150/367)   \rResolving deltas:  41% (153/367)   \rResolving deltas:  55% (202/367)   \rResolving deltas:  61% (224/367)   \rResolving deltas:  63% (234/367)   \rResolving deltas:  64% (237/367)   \rResolving deltas:  67% (246/367)   \rResolving deltas:  74% (274/367)   \rResolving deltas:  75% (278/367)   \rResolving deltas:  78% (288/367)   \rResolving deltas:  83% (306/367)   \rResolving deltas:  85% (315/367)   \rResolving deltas:  87% (320/367)   \rResolving deltas:  88% (323/367)   \rResolving deltas:  91% (336/367)   \rResolving deltas:  92% (340/367)   \rResolving deltas:  94% (347/367)   \rResolving deltas:  95% (352/367)   \rResolving deltas:  97% (357/367)   \rResolving deltas:  99% (364/367)   \rResolving deltas: 100% (367/367)   \rResolving deltas: 100% (367/367), done.\n",
            "Uninstalling seqdataloader-0.126:\n",
            "  Would remove:\n",
            "    /usr/local/bin/genomewide_labels\n",
            "    /usr/local/lib/python3.6/dist-packages/seqdataloader-0.126.dist-info/*\n",
            "    /usr/local/lib/python3.6/dist-packages/seqdataloader/*\n",
            "Proceed (y/n)? y\n",
            "  Successfully uninstalled seqdataloader-0.126\n",
            "Processing ./seqdataloader\n",
            "Requirement already satisfied: numpy>=1.15 in /usr/local/lib/python3.6/dist-packages (from seqdataloader==0.126) (1.16.4)\n",
            "Requirement already satisfied: pandas>=0.23.4 in /usr/local/lib/python3.6/dist-packages (from seqdataloader==0.126) (0.24.2)\n",
            "Requirement already satisfied: cython>=0.27.3 in /usr/local/lib/python3.6/dist-packages (from seqdataloader==0.126) (0.29.10)\n",
            "Requirement already satisfied: deeptools>=3.0.1 in /usr/local/lib/python3.6/dist-packages (from seqdataloader==0.126) (3.3.0)\n",
            "Requirement already satisfied: pybedtools>=0.7 in /usr/local/lib/python3.6/dist-packages (from seqdataloader==0.126) (0.8.0)\n",
            "Requirement already satisfied: pyBigWig>=0.3.2 in /usr/local/lib/python3.6/dist-packages (from seqdataloader==0.126) (0.3.16)\n",
            "Requirement already satisfied: pyfaidx in /usr/local/lib/python3.6/dist-packages (from seqdataloader==0.126) (0.5.5.2)\n",
            "Requirement already satisfied: python-dateutil>=2.5.0 in /usr/local/lib/python3.6/dist-packages (from pandas>=0.23.4->seqdataloader==0.126) (2.5.3)\n",
            "Requirement already satisfied: pytz>=2011k in /usr/local/lib/python3.6/dist-packages (from pandas>=0.23.4->seqdataloader==0.126) (2018.9)\n",
            "Requirement already satisfied: pysam>=0.14.0 in /usr/local/lib/python3.6/dist-packages (from deeptools>=3.0.1->seqdataloader==0.126) (0.15.2)\n",
            "Requirement already satisfied: plotly>=2.0.0 in /usr/local/lib/python3.6/dist-packages (from deeptools>=3.0.1->seqdataloader==0.126) (3.6.1)\n",
            "Requirement already satisfied: numpydoc>=0.5 in /usr/local/lib/python3.6/dist-packages (from deeptools>=3.0.1->seqdataloader==0.126) (0.9.1)\n",
            "Requirement already satisfied: scipy>=0.17.0 in /usr/local/lib/python3.6/dist-packages (from deeptools>=3.0.1->seqdataloader==0.126) (1.3.0)\n",
            "Requirement already satisfied: py2bit>=0.2.0 in /usr/local/lib/python3.6/dist-packages (from deeptools>=3.0.1->seqdataloader==0.126) (0.3.0)\n",
            "Requirement already satisfied: deeptoolsintervals>=0.1.8 in /usr/local/lib/python3.6/dist-packages (from deeptools>=3.0.1->seqdataloader==0.126) (0.1.8)\n",
            "Requirement already satisfied: matplotlib>=2.1.2 in /usr/local/lib/python3.6/dist-packages (from deeptools>=3.0.1->seqdataloader==0.126) (3.0.3)\n",
            "Requirement already satisfied: six in /usr/local/lib/python3.6/dist-packages (from pybedtools>=0.7->seqdataloader==0.126) (1.12.0)\n",
            "Requirement already satisfied: setuptools>=0.7 in /usr/local/lib/python3.6/dist-packages (from pyfaidx->seqdataloader==0.126) (41.0.1)\n",
            "Requirement already satisfied: nbformat>=4.2 in /usr/local/lib/python3.6/dist-packages (from plotly>=2.0.0->deeptools>=3.0.1->seqdataloader==0.126) (4.4.0)\n",
            "Requirement already satisfied: retrying>=1.3.3 in /usr/local/lib/python3.6/dist-packages (from plotly>=2.0.0->deeptools>=3.0.1->seqdataloader==0.126) (1.3.3)\n",
            "Requirement already satisfied: decorator>=4.0.6 in /usr/local/lib/python3.6/dist-packages (from plotly>=2.0.0->deeptools>=3.0.1->seqdataloader==0.126) (4.4.0)\n",
            "Requirement already satisfied: requests in /usr/local/lib/python3.6/dist-packages (from plotly>=2.0.0->deeptools>=3.0.1->seqdataloader==0.126) (2.21.0)\n",
            "Requirement already satisfied: Jinja2>=2.3 in /usr/local/lib/python3.6/dist-packages (from numpydoc>=0.5->deeptools>=3.0.1->seqdataloader==0.126) (2.10.1)\n",
            "Requirement already satisfied: sphinx>=1.6.5 in /usr/local/lib/python3.6/dist-packages (from numpydoc>=0.5->deeptools>=3.0.1->seqdataloader==0.126) (1.8.5)\n",
            "Requirement already satisfied: kiwisolver>=1.0.1 in /usr/local/lib/python3.6/dist-packages (from matplotlib>=2.1.2->deeptools>=3.0.1->seqdataloader==0.126) (1.1.0)\n",
            "Requirement already satisfied: cycler>=0.10 in /usr/local/lib/python3.6/dist-packages (from matplotlib>=2.1.2->deeptools>=3.0.1->seqdataloader==0.126) (0.10.0)\n",
            "Requirement already satisfied: pyparsing!=2.0.4,!=2.1.2,!=2.1.6,>=2.0.1 in /usr/local/lib/python3.6/dist-packages (from matplotlib>=2.1.2->deeptools>=3.0.1->seqdataloader==0.126) (2.4.0)\n",
            "Requirement already satisfied: traitlets>=4.1 in /usr/local/lib/python3.6/dist-packages (from nbformat>=4.2->plotly>=2.0.0->deeptools>=3.0.1->seqdataloader==0.126) (4.3.2)\n",
            "Requirement already satisfied: ipython-genutils in /usr/local/lib/python3.6/dist-packages (from nbformat>=4.2->plotly>=2.0.0->deeptools>=3.0.1->seqdataloader==0.126) (0.2.0)\n",
            "Requirement already satisfied: jupyter-core in /usr/local/lib/python3.6/dist-packages (from nbformat>=4.2->plotly>=2.0.0->deeptools>=3.0.1->seqdataloader==0.126) (4.5.0)\n",
            "Requirement already satisfied: jsonschema!=2.5.0,>=2.4 in /usr/local/lib/python3.6/dist-packages (from nbformat>=4.2->plotly>=2.0.0->deeptools>=3.0.1->seqdataloader==0.126) (2.6.0)\n",
            "Requirement already satisfied: chardet<3.1.0,>=3.0.2 in /usr/local/lib/python3.6/dist-packages (from requests->plotly>=2.0.0->deeptools>=3.0.1->seqdataloader==0.126) (3.0.4)\n",
            "Requirement already satisfied: urllib3<1.25,>=1.21.1 in /usr/local/lib/python3.6/dist-packages (from requests->plotly>=2.0.0->deeptools>=3.0.1->seqdataloader==0.126) (1.24.3)\n",
            "Requirement already satisfied: idna<2.9,>=2.5 in /usr/local/lib/python3.6/dist-packages (from requests->plotly>=2.0.0->deeptools>=3.0.1->seqdataloader==0.126) (2.8)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.6/dist-packages (from requests->plotly>=2.0.0->deeptools>=3.0.1->seqdataloader==0.126) (2019.6.16)\n",
            "Requirement already satisfied: MarkupSafe>=0.23 in /usr/local/lib/python3.6/dist-packages (from Jinja2>=2.3->numpydoc>=0.5->deeptools>=3.0.1->seqdataloader==0.126) (1.1.1)\n",
            "Requirement already satisfied: Pygments>=2.0 in /usr/local/lib/python3.6/dist-packages (from sphinx>=1.6.5->numpydoc>=0.5->deeptools>=3.0.1->seqdataloader==0.126) (2.1.3)\n",
            "Requirement already satisfied: alabaster<0.8,>=0.7 in /usr/local/lib/python3.6/dist-packages (from sphinx>=1.6.5->numpydoc>=0.5->deeptools>=3.0.1->seqdataloader==0.126) (0.7.12)\n",
            "Requirement already satisfied: packaging in /usr/local/lib/python3.6/dist-packages (from sphinx>=1.6.5->numpydoc>=0.5->deeptools>=3.0.1->seqdataloader==0.126) (19.0)\n",
            "Requirement already satisfied: sphinxcontrib-websupport in /usr/local/lib/python3.6/dist-packages (from sphinx>=1.6.5->numpydoc>=0.5->deeptools>=3.0.1->seqdataloader==0.126) (1.1.2)\n",
            "Requirement already satisfied: snowballstemmer>=1.1 in /usr/local/lib/python3.6/dist-packages (from sphinx>=1.6.5->numpydoc>=0.5->deeptools>=3.0.1->seqdataloader==0.126) (1.2.1)\n",
            "Requirement already satisfied: docutils>=0.11 in /usr/local/lib/python3.6/dist-packages (from sphinx>=1.6.5->numpydoc>=0.5->deeptools>=3.0.1->seqdataloader==0.126) (0.14)\n",
            "Requirement already satisfied: babel!=2.0,>=1.3 in /usr/local/lib/python3.6/dist-packages (from sphinx>=1.6.5->numpydoc>=0.5->deeptools>=3.0.1->seqdataloader==0.126) (2.7.0)\n",
            "Requirement already satisfied: imagesize in /usr/local/lib/python3.6/dist-packages (from sphinx>=1.6.5->numpydoc>=0.5->deeptools>=3.0.1->seqdataloader==0.126) (1.1.0)\n",
            "Building wheels for collected packages: seqdataloader\n",
            "  Building wheel for seqdataloader (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Stored in directory: /tmp/pip-ephem-wheel-cache-eavjuugh/wheels/c2/db/13/112d41662f69fb8c7986c218293570cc1550fc21eed966e31b\n",
            "Successfully built seqdataloader\n",
            "Installing collected packages: seqdataloader\n",
            "Successfully installed seqdataloader-0.126\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "KTCqs6AEPASm",
        "colab_type": "code",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "outputId": "0813de31-96f1-4566-fc3d-9a3657dc464e"
      },
      "source": [
        "from collections import OrderedDict\n",
        "from seqdataloader.batchproducers import coordbased\n",
        "from seqdataloader.batchproducers.coordbased import coordstovals\n",
        "from seqdataloader.batchproducers.coordbased import coordbatchproducers\n",
        "from seqdataloader.batchproducers.coordbased import coordbatchtransformers\n",
        "import keras\n",
        "import keras.layers as kl\n",
        "import tensorflow as tf\n",
        "import tensorflow_probability as tfp\n",
        "import numpy as np\n",
        "from seqdataloader.batchproducers import coordbased\n",
        "from seqdataloader.batchproducers.coordbased import coordstovals\n",
        "from seqdataloader.batchproducers.coordbased import coordbatchproducers\n",
        "from seqdataloader.batchproducers.coordbased import coordbatchtransformers\n",
        "\n",
        "#Written by Žiga Avsec\n",
        "def multinomial_nll(true_counts, logits):\n",
        "    \"\"\"Compute the multinomial negative log-likelihood\n",
        "    Args:\n",
        "      true_counts: observed count values\n",
        "      logits: predicted logit values\n",
        "    \"\"\"\n",
        "    counts_per_example = tf.reduce_sum(true_counts, axis=-1)\n",
        "    dist = tfp.distributions.Multinomial(total_count=counts_per_example,\n",
        "                                         logits=logits)\n",
        "    return (-tf.reduce_sum(dist.log_prob(true_counts)) / \n",
        "            tf.to_float(tf.shape(true_counts)[0]))\n",
        "\n",
        "#Written by Žiga Avsec\n",
        "class MultichannelMultinomialNLL(object):\n",
        "    def __init__(self, n):\n",
        "        self.__name__ = \"MultichannelMultinomialNLL\"\n",
        "        self.n = n\n",
        "\n",
        "    def __call__(self, true_counts, logits):\n",
        "        for i in range(self.n):\n",
        "            loss = multinomial_nll(true_counts[..., i], logits[..., i])\n",
        "            if i == 0:\n",
        "                total = loss\n",
        "            else:\n",
        "                total += loss\n",
        "        return total\n",
        "\n",
        "    def get_config(self):\n",
        "        return {\"n\": self.n}\n",
        "\n",
        "#If we want to avoid zero-padding, then the size of the output predictions\n",
        "# will depend on the size of the input sequence supplied. We define the\n",
        "# API for an AbstractProfileModel class which returns the length of the\n",
        "# output profile in addition to returning the model, given information\n",
        "# on the size of the input sequence and the model parameters.\n",
        "class AbstractProfileModel(object):\n",
        "    \n",
        "    def get_output_profile_len(self):\n",
        "        raise NotImplementedError()\n",
        "  \n",
        "    def get_model(self):\n",
        "        raise NotImplementedError()\n",
        "  \n",
        "#The architecture by Žiga Avsec involves residual connections, which means\n",
        "# that the layers being added together in an elementwise fashion need\n",
        "# to have the same dimensions. To achieve this without zero-padding, we\n",
        "# have to trim away the flanks of earlier convolutional layers. That\n",
        "# is what this function is meant to do. (Note that the original BP-net\n",
        "# architecture zero-pads; this is a modification to avoid the zero\n",
        "# padding and use information on actual sequence instead)\n",
        "def trim_flanks_of_conv_layer(conv_layer, output_len, width_to_trim, filters):\n",
        "    layer = keras.layers.Lambda(\n",
        "        lambda x: x[:,\n",
        "          int(0.5*(width_to_trim)):-(width_to_trim-int(0.5*(width_to_trim)))],\n",
        "        output_shape=(output_len, filters))(conv_layer)\n",
        "    return layer\n",
        "        \n",
        "#This model architecture is based on BP-Net by Žiga Avsec\n",
        "# https://drive.google.com/file/d/1kg6Ic0-FvJtVUva9Mh3FPnOAHJcN6VB-/view\n",
        "#It has been modified for this specific use-case.\n",
        "class BPnetArch(AbstractProfileModel):   \n",
        "\n",
        "    def __init__(self, input_seq_len, c_task_weight, filters,\n",
        "                       n_dil_layers, conv1_kernel_size,\n",
        "                       dil_kernel_size,\n",
        "                       outconv_kernel_size, lr):\n",
        "        self.input_seq_len = input_seq_len\n",
        "        self.c_task_weight = c_task_weight\n",
        "        self.filters = filters\n",
        "        self.n_dil_layers = n_dil_layers\n",
        "        self.conv1_kernel_size = conv1_kernel_size\n",
        "        self.dil_kernel_size = dil_kernel_size\n",
        "        self.outconv_kernel_size = outconv_kernel_size\n",
        "        self.lr = lr\n",
        "    \n",
        "    def get_embedding_len(self):\n",
        "        embedding_len = self.input_seq_len\n",
        "        embedding_len -= (self.conv1_kernel_size-1)     \n",
        "        for i in range(1, self.n_dil_layers+1):\n",
        "            dilation_rate = (2**i)\n",
        "            embedding_len -= dilation_rate*(self.dil_kernel_size-1)\n",
        "        return embedding_len\n",
        "    \n",
        "    def get_output_profile_len(self):\n",
        "        embedding_len = self.get_embedding_len()\n",
        "        out_profile_len = embedding_len - (self.outconv_kernel_size - 1)\n",
        "        return out_profile_len\n",
        "    \n",
        "    def get_keras_model(self):\n",
        "      \n",
        "        out_pred_len = self.get_output_profile_len()\n",
        "        \n",
        "        #'inp' is the one-hot encoded DNA sequence input\n",
        "        inp = kl.Input(shape=(self.input_seq_len, 4), name='sequence')\n",
        "        first_conv = kl.Conv1D(filters=self.filters,\n",
        "                               kernel_size=self.conv1_kernel_size,\n",
        "                               padding='valid',\n",
        "                               activation='relu')(inp)\n",
        "        #Need to keep track of the layer size for trimming purposes when\n",
        "        # we get to the residual connections.\n",
        "        curr_layer_size = self.input_seq_len - (self.conv1_kernel_size-1)\n",
        "        \n",
        "        #Define input layers for the control tracks - both counts and profile\n",
        "        #Dimension is '1' for the ChIP-seq control counts because the positive\n",
        "        # and negative strands are added together\n",
        "        #Dimension is '3' for the CUTNRUN counts because we have all fragment\n",
        "        # counts, counts for frag len < 120bp and counts for frag len > 150bp.\n",
        "        chipseq_control_counts_input = kl.Input(\n",
        "            shape=(1,), name=\"ChIPseq.control.logcount\")\n",
        "        cutnrun_control_standard_counts_input = kl.Input(\n",
        "            shape=(3,), name=\"CUTNRUN.control.standard.logcount\")\n",
        "        cutnrun_control_auto_counts_input = kl.Input(\n",
        "            shape=(3,), name=\"CUTNRUN.control.auto.logcount\")\n",
        "        #Second dimension for the profile tasks are twice those for the\n",
        "        # respective counts tasks because in addition to the original profile,\n",
        "        # we supply the profile smoothed by 50bp.\n",
        "        chipseq_control_profile_input = kl.Input(\n",
        "            shape=(out_pred_len, 2), name=\"ChIPseq.control.profile\")\n",
        "        cutnrun_control_standard_profile_input = kl.Input(\n",
        "            shape=(out_pred_len, 6), name=\"CUTNRUN.control.standard.profile\")\n",
        "        cutnrun_control_auto_profile_input = kl.Input(\n",
        "            shape=(out_pred_len, 6), name=\"CUTNRUN.control.auto.profile\")\n",
        "        \n",
        "        #Gather together all the tensors representing the model inputs\n",
        "        model_inputs = [\n",
        "            inp,\n",
        "            chipseq_control_counts_input,\n",
        "            cutnrun_control_standard_counts_input,\n",
        "            cutnrun_control_auto_counts_input,\n",
        "            chipseq_control_profile_input,\n",
        "            cutnrun_control_standard_profile_input,\n",
        "            cutnrun_control_auto_profile_input]\n",
        "        \n",
        "        #Prepare the stack of dilated convolutions with residual connections\n",
        "        prev_layers = [first_conv]\n",
        "        for i in range(1, self.n_dil_layers + 1):\n",
        "          dilation_rate = 2**i\n",
        "          if i == 1:\n",
        "              prev_sum = first_conv\n",
        "          else:\n",
        "              prev_sum = kl.merge.Average()(prev_layers)\n",
        "          conv_output = kl.Conv1D(filters=self.filters,\n",
        "                                  kernel_size=self.dil_kernel_size,\n",
        "                                  padding='valid',\n",
        "                                  activation='relu',\n",
        "                                  dilation_rate=dilation_rate)(prev_sum)          \n",
        "          width_to_trim = dilation_rate*(self.dil_kernel_size-1)\n",
        "          curr_layer_size = (curr_layer_size - width_to_trim)\n",
        "          prev_layers = [trim_flanks_of_conv_layer(\n",
        "              conv_layer=x, output_len=curr_layer_size,\n",
        "              width_to_trim=width_to_trim, filters=self.filters)\n",
        "              for x in prev_layers]\n",
        "          prev_layers.append(conv_output)\n",
        "\n",
        "        #combined_conv is the final output of the dilated convolution stack\n",
        "        # note that I (Avanti Shrikumar)\n",
        "        # use Average instead of Add. The reasoning was that Add may cause\n",
        "        # exploding variances in the activations.\n",
        "        combined_conv = kl.merge.Average()(prev_layers)\n",
        "\n",
        "        #gap = GlobalAveragePooling. This layer is used as input for the\n",
        "        # counts prediction tasks.\n",
        "        gap_combined_conv = kl.GlobalAvgPool1D()(combined_conv)\n",
        "        \n",
        "        lossarr = []\n",
        "        lossweightsarr = []\n",
        "        model_outputs = []\n",
        "        \n",
        "        #Define the output layers for the counts prediction tasks\n",
        "        for countouttaskname, numunits, countcontrolinp in [\n",
        "          (\"ChIPseq.POU5F1.logcount\", 2, chipseq_control_counts_input),\n",
        "          (\"ChIPseq.NANOG.logcount\", 2, chipseq_control_counts_input),\n",
        "          (\"CUTNRUN.POU5F1.logcount\", 3, cutnrun_control_standard_counts_input),\n",
        "          (\"CUTNRUN.NANOG.logcount\", 3, cutnrun_control_auto_counts_input)]:\n",
        "          count_out = kl.Dense(units=numunits,\n",
        "                                 name=countouttaskname)(\n",
        "              kl.concatenate([gap_combined_conv, countcontrolinp], axis=-1))\n",
        "          model_outputs.append(count_out)\n",
        "          lossarr.append('mse')\n",
        "          lossweightsarr.append(self.c_task_weight)\n",
        "\n",
        "        #Define the output layers for the profile prediction tasks\n",
        "        for profileouttaskname, numunits, profilecontrolinp in [\n",
        "          (\"ChIPseq.POU5F1.profile\", 2, chipseq_control_profile_input),\n",
        "          (\"ChIPseq.NANOG.profile\", 2, chipseq_control_profile_input),\n",
        "          (\"CUTNRUN.POU5F1.profile\", 3, cutnrun_control_standard_profile_input),\n",
        "          (\"CUTNRUN.NANOG.profile\", 3, cutnrun_control_auto_profile_input)]:     \n",
        "          profile_out_precontrol = kl.Conv1D(\n",
        "                                 filters=numunits,\n",
        "                                 kernel_size=self.outconv_kernel_size,\n",
        "                                 padding='valid')(combined_conv)\n",
        "          profile_out = kl.Conv1D(\n",
        "              filters=numunits, kernel_size=1, name=profileouttaskname)(\n",
        "                      kl.concatenate([profile_out_precontrol,\n",
        "                                      profilecontrolinp], axis=-1))\n",
        "          model_outputs.append(profile_out)\n",
        "          lossarr.append(MultichannelMultinomialNLL(numunits)) \n",
        "          #We downweight the loss by the number of channels because, if you\n",
        "          # read the code for MultichannelMultinomialNLL, you'll see that\n",
        "          # the loss for different channels is added together;\n",
        "          # but I (Avanti Shrikumar) didn't want to implicitly upweight\n",
        "          # the prediction tasks that happen\n",
        "          # to have more channels, hence this downweighting.\n",
        "          lossweightsarr.append(1.0/numunits)\n",
        "\n",
        "        #Compile the model and return it\n",
        "        model = keras.models.Model(inputs=model_inputs, outputs=model_outputs)\n",
        "        model.compile(keras.optimizers.Adam(lr=self.lr),\n",
        "                      loss=lossarr,\n",
        "                      loss_weights=lossweightsarr)\n",
        "        return model\n",
        "\n",
        "seq_len = 1000\n",
        "modelwrapper = BPnetArch(\n",
        "    input_seq_len=seq_len, c_task_weight=50,\n",
        "    filters=64, n_dil_layers=6,\n",
        "    conv1_kernel_size=21,\n",
        "    dil_kernel_size=3,\n",
        "    outconv_kernel_size=75,\n",
        "    lr=0.001)\n",
        "out_pred_len = modelwrapper.get_output_profile_len()\n",
        "print(out_pred_len, seq_len-out_pred_len)\n",
        "\n",
        "#the code below is used to prepare instances of keras.utils.Sequence that\n",
        "# can be supplied to model.fit_generator(...)\n",
        "#Note that we log-transform our counts using np.log(counts+1)\n",
        "#Also note that the profiles for the control are smoothed by windows of\n",
        "# size 1 and 50 (smoothing by a window of size 1 just returns\n",
        "# the original profile)\n",
        "inputs_coordstovals = coordstovals.core.CoordsToValsJoiner(\n",
        "    coordstovals_list=[\n",
        "      coordbased.coordstovals.fasta.PyfaidxCoordsToVals(\n",
        "        genome_fasta_path=\"hg38.genome.fa\",\n",
        "        mode_name=\"sequence\",\n",
        "        center_size_to_use=seq_len),\n",
        "      coordstovals.bigwig.MultiTrackCountsAndProfile(\n",
        "        counts_mode_name=\"ChIPseq.control.logcount\",\n",
        "        profile_mode_name=\"ChIPseq.control.profile\",\n",
        "        bigwig_paths=[\"ChIPseq.control.bigWig\"],\n",
        "        counts_and_profiles_transformer=\n",
        "          coordstovals.bigwig.LogCountsPlusOne().chain(\n",
        "            coordstovals.bigwig.SmoothProfiles(smoothing_windows=[1,50])),        \n",
        "        center_size_to_use=out_pred_len),\n",
        "     coordstovals.bigwig.MultiTrackCountsAndProfile(\n",
        "        counts_mode_name=\"CUTNRUN.control.auto.logcount\",\n",
        "        profile_mode_name=\"CUTNRUN.control.auto.profile\",\n",
        "        bigwig_paths=[\"CUTNRUN.IgG.merged.auto.5p.counts.bigWig\",\n",
        "                      \"CUTNRUN.IgG.merged.auto.maxFL120bp.5p.counts.bigWig\",\n",
        "                      \"CUTNRUN.IgG.merged.auto.minFL150bp.5p.counts.bigWig\"],\n",
        "        counts_and_profiles_transformer=\n",
        "          coordstovals.bigwig.LogCountsPlusOne().chain(\n",
        "            coordstovals.bigwig.SmoothProfiles(smoothing_windows=[1,50])),\n",
        "        center_size_to_use=out_pred_len),\n",
        "     coordstovals.bigwig.MultiTrackCountsAndProfile(\n",
        "        counts_mode_name=\"CUTNRUN.control.standard.logcount\",\n",
        "        profile_mode_name=\"CUTNRUN.control.standard.profile\",\n",
        "        bigwig_paths=[\"CUTNRUN.IgG.standard.5p.counts.bigWig\",\n",
        "                      \"CUTNRUN.IgG.standard.maxFL120bp.5p.counts.bigWig\",\n",
        "                      \"CUTNRUN.IgG.standard.minFL150bp.5p.counts.bigWig\"],\n",
        "        counts_and_profiles_transformer=\n",
        "          coordstovals.bigwig.LogCountsPlusOne().chain(\n",
        "            coordstovals.bigwig.SmoothProfiles(smoothing_windows=[1,50])),       \n",
        "        center_size_to_use=out_pred_len)\n",
        "    ])\n",
        "targets_coordstovals = coordstovals.core.CoordsToValsJoiner(\n",
        "    coordstovals_list=[\n",
        "      coordstovals.bigwig.PosAndNegSeparateLogCounts(\n",
        "        counts_mode_name=\"ChIPseq.POU5F1.logcount\",\n",
        "        profile_mode_name=\"ChIPseq.POU5F1.profile\",\n",
        "        pos_strand_bigwig_path=\"ChIPseq.POU5F1.merged.5p.counts.plus.bigWig\",\n",
        "        neg_strand_bigwig_path=\"ChIPseq.POU5F1.merged.5p.counts.minus.bigWig\",\n",
        "        center_size_to_use=out_pred_len),\n",
        "      coordstovals.bigwig.PosAndNegSeparateLogCounts(\n",
        "        counts_mode_name=\"ChIPseq.NANOG.logcount\",\n",
        "        profile_mode_name=\"ChIPseq.NANOG.profile\",\n",
        "        pos_strand_bigwig_path=\"ChIPseq.NANOG.merged.5p.counts.plus.bigWig\",\n",
        "        neg_strand_bigwig_path=\"ChIPseq.NANOG.merged.5p.counts.minus.bigWig\",\n",
        "        center_size_to_use=out_pred_len),\n",
        "      coordstovals.bigwig.MultiTrackCountsAndProfile(\n",
        "        counts_mode_name=\"CUTNRUN.POU5F1.logcount\",\n",
        "        profile_mode_name=\"CUTNRUN.POU5F1.profile\",\n",
        "        bigwig_paths=[\"CUTNRUN.POU5F1.5p.counts.bigWig\",\n",
        "                      \"CUTNRUN.POU5F1.maxFL120bp.5p.counts.bigWig\",\n",
        "                      \"CUTNRUN.POU5F1.minFL150bp.5p.counts.bigWig\"],\n",
        "        counts_and_profiles_transformer=coordstovals.bigwig.LogCountsPlusOne(),\n",
        "        center_size_to_use=out_pred_len),\n",
        "      coordstovals.bigwig.MultiTrackCountsAndProfile(\n",
        "        counts_mode_name=\"CUTNRUN.NANOG.logcount\",\n",
        "        profile_mode_name=\"CUTNRUN.NANOG.profile\",\n",
        "        bigwig_paths=[\"CUTNRUN.NANOG.5p.counts.bigWig\",\n",
        "                      \"CUTNRUN.NANOG.maxFL120bp.5p.counts.bigWig\",\n",
        "                      \"CUTNRUN.NANOG.minFL150bp.5p.counts.bigWig\"],\n",
        "        counts_and_profiles_transformer=coordstovals.bigwig.LogCountsPlusOne(),\n",
        "        center_size_to_use=out_pred_len)\n",
        "    ])\n",
        "keras_train_batch_generator = coordbased.core.KerasBatchGenerator(\n",
        "  coordsbatch_producer=coordbatchproducers.SimpleCoordsBatchProducer(\n",
        "      bed_file=\"train_2k_around_summits.bed.gz\",\n",
        "      coord_batch_transformer=\n",
        "          coordbatchtransformers.ReverseComplementAugmenter().chain(\n",
        "          coordbatchtransformers.UniformJitter(\n",
        "              maxshift=200, chromsizes_file=\"hg38.chrom.sizes\")),\n",
        "      batch_size=64,\n",
        "      shuffle_before_epoch=True, \n",
        "      seed=1234),\n",
        "  inputs_coordstovals=inputs_coordstovals,\n",
        "  targets_coordstovals=targets_coordstovals)\n",
        "keras_valid_batch_generator = coordbased.core.KerasBatchGenerator(\n",
        "  coordsbatch_producer=coordbatchproducers.SimpleCoordsBatchProducer(\n",
        "            bed_file=\"valid_2k_around_summits.bed.gz\",\n",
        "            batch_size=64,\n",
        "            shuffle_before_epoch=False, \n",
        "            seed=1234),\n",
        "  inputs_coordstovals=inputs_coordstovals,\n",
        "  targets_coordstovals=targets_coordstovals)\n",
        "\n",
        "thebatch = keras_train_batch_generator[0]\n",
        "for tupleidx,tupleentry in enumerate(thebatch):\n",
        "  print(\"Tuple entry\",tupleidx)\n",
        "  for key in tupleentry:\n",
        "    print(key, tupleentry[key].shape)\n",
        "    \n",
        "model = modelwrapper.get_keras_model()\n",
        "print(model.summary())\n",
        "early_stopping_callback = keras.callbacks.EarlyStopping(\n",
        "                            patience=3, restore_best_weights=True)\n",
        "loss_history = model.fit_generator(keras_train_batch_generator,\n",
        "                    epochs=200,\n",
        "                    validation_data=keras_valid_batch_generator,\n",
        "                    callbacks=[early_stopping_callback])\n",
        "model.set_weights(early_stopping_callback.best_weights)"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "654 346\n",
            "Tuple entry 0\n",
            "sequence (128, 1000, 4)\n",
            "ChIPseq.control.logcount (128, 1)\n",
            "ChIPseq.control.profile (128, 654, 2)\n",
            "CUTNRUN.control.auto.logcount (128, 3)\n",
            "CUTNRUN.control.auto.profile (128, 654, 6)\n",
            "CUTNRUN.control.standard.logcount (128, 3)\n",
            "CUTNRUN.control.standard.profile (128, 654, 6)\n",
            "Tuple entry 1\n",
            "ChIPseq.POU5F1.logcount (128, 2)\n",
            "ChIPseq.POU5F1.profile (128, 654, 2)\n",
            "ChIPseq.NANOG.logcount (128, 2)\n",
            "ChIPseq.NANOG.profile (128, 654, 2)\n",
            "CUTNRUN.POU5F1.logcount (128, 3)\n",
            "CUTNRUN.POU5F1.profile (128, 654, 3)\n",
            "CUTNRUN.NANOG.logcount (128, 3)\n",
            "CUTNRUN.NANOG.profile (128, 654, 3)\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "stream",
          "text": [
            "W0707 14:37:02.203984 140481616963456 deprecation.py:323] From <ipython-input-10-8d48c4eb8b4e>:26: to_float (from tensorflow.python.ops.math_ops) is deprecated and will be removed in a future version.\n",
            "Instructions for updating:\n",
            "Use `tf.cast` instead.\n"
          ],
          "name": "stderr"
        },
        {
          "output_type": "stream",
          "text": [
            "__________________________________________________________________________________________________\n",
            "Layer (type)                    Output Shape         Param #     Connected to                     \n",
            "==================================================================================================\n",
            "sequence (InputLayer)           (None, 1000, 4)      0                                            \n",
            "__________________________________________________________________________________________________\n",
            "conv1d_37 (Conv1D)              (None, 980, 64)      5440        sequence[0][0]                   \n",
            "__________________________________________________________________________________________________\n",
            "lambda_86 (Lambda)              (None, 976, 64)      0           conv1d_37[0][0]                  \n",
            "__________________________________________________________________________________________________\n",
            "conv1d_38 (Conv1D)              (None, 976, 64)      12352       conv1d_37[0][0]                  \n",
            "__________________________________________________________________________________________________\n",
            "average_26 (Average)            (None, 976, 64)      0           lambda_86[0][0]                  \n",
            "                                                                 conv1d_38[0][0]                  \n",
            "__________________________________________________________________________________________________\n",
            "lambda_87 (Lambda)              (None, 968, 64)      0           lambda_86[0][0]                  \n",
            "__________________________________________________________________________________________________\n",
            "lambda_88 (Lambda)              (None, 968, 64)      0           conv1d_38[0][0]                  \n",
            "__________________________________________________________________________________________________\n",
            "conv1d_39 (Conv1D)              (None, 968, 64)      12352       average_26[0][0]                 \n",
            "__________________________________________________________________________________________________\n",
            "average_27 (Average)            (None, 968, 64)      0           lambda_87[0][0]                  \n",
            "                                                                 lambda_88[0][0]                  \n",
            "                                                                 conv1d_39[0][0]                  \n",
            "__________________________________________________________________________________________________\n",
            "lambda_89 (Lambda)              (None, 952, 64)      0           lambda_87[0][0]                  \n",
            "__________________________________________________________________________________________________\n",
            "lambda_90 (Lambda)              (None, 952, 64)      0           lambda_88[0][0]                  \n",
            "__________________________________________________________________________________________________\n",
            "lambda_91 (Lambda)              (None, 952, 64)      0           conv1d_39[0][0]                  \n",
            "__________________________________________________________________________________________________\n",
            "conv1d_40 (Conv1D)              (None, 952, 64)      12352       average_27[0][0]                 \n",
            "__________________________________________________________________________________________________\n",
            "average_28 (Average)            (None, 952, 64)      0           lambda_89[0][0]                  \n",
            "                                                                 lambda_90[0][0]                  \n",
            "                                                                 lambda_91[0][0]                  \n",
            "                                                                 conv1d_40[0][0]                  \n",
            "__________________________________________________________________________________________________\n",
            "lambda_92 (Lambda)              (None, 920, 64)      0           lambda_89[0][0]                  \n",
            "__________________________________________________________________________________________________\n",
            "lambda_93 (Lambda)              (None, 920, 64)      0           lambda_90[0][0]                  \n",
            "__________________________________________________________________________________________________\n",
            "lambda_94 (Lambda)              (None, 920, 64)      0           lambda_91[0][0]                  \n",
            "__________________________________________________________________________________________________\n",
            "lambda_95 (Lambda)              (None, 920, 64)      0           conv1d_40[0][0]                  \n",
            "__________________________________________________________________________________________________\n",
            "conv1d_41 (Conv1D)              (None, 920, 64)      12352       average_28[0][0]                 \n",
            "__________________________________________________________________________________________________\n",
            "average_29 (Average)            (None, 920, 64)      0           lambda_92[0][0]                  \n",
            "                                                                 lambda_93[0][0]                  \n",
            "                                                                 lambda_94[0][0]                  \n",
            "                                                                 lambda_95[0][0]                  \n",
            "                                                                 conv1d_41[0][0]                  \n",
            "__________________________________________________________________________________________________\n",
            "lambda_96 (Lambda)              (None, 856, 64)      0           lambda_92[0][0]                  \n",
            "__________________________________________________________________________________________________\n",
            "lambda_97 (Lambda)              (None, 856, 64)      0           lambda_93[0][0]                  \n",
            "__________________________________________________________________________________________________\n",
            "lambda_98 (Lambda)              (None, 856, 64)      0           lambda_94[0][0]                  \n",
            "__________________________________________________________________________________________________\n",
            "lambda_99 (Lambda)              (None, 856, 64)      0           lambda_95[0][0]                  \n",
            "__________________________________________________________________________________________________\n",
            "lambda_100 (Lambda)             (None, 856, 64)      0           conv1d_41[0][0]                  \n",
            "__________________________________________________________________________________________________\n",
            "conv1d_42 (Conv1D)              (None, 856, 64)      12352       average_29[0][0]                 \n",
            "__________________________________________________________________________________________________\n",
            "average_30 (Average)            (None, 856, 64)      0           lambda_96[0][0]                  \n",
            "                                                                 lambda_97[0][0]                  \n",
            "                                                                 lambda_98[0][0]                  \n",
            "                                                                 lambda_99[0][0]                  \n",
            "                                                                 lambda_100[0][0]                 \n",
            "                                                                 conv1d_42[0][0]                  \n",
            "__________________________________________________________________________________________________\n",
            "lambda_101 (Lambda)             (None, 728, 64)      0           lambda_96[0][0]                  \n",
            "__________________________________________________________________________________________________\n",
            "lambda_102 (Lambda)             (None, 728, 64)      0           lambda_97[0][0]                  \n",
            "__________________________________________________________________________________________________\n",
            "lambda_103 (Lambda)             (None, 728, 64)      0           lambda_98[0][0]                  \n",
            "__________________________________________________________________________________________________\n",
            "lambda_104 (Lambda)             (None, 728, 64)      0           lambda_99[0][0]                  \n",
            "__________________________________________________________________________________________________\n",
            "lambda_105 (Lambda)             (None, 728, 64)      0           lambda_100[0][0]                 \n",
            "__________________________________________________________________________________________________\n",
            "lambda_106 (Lambda)             (None, 728, 64)      0           conv1d_42[0][0]                  \n",
            "__________________________________________________________________________________________________\n",
            "conv1d_43 (Conv1D)              (None, 728, 64)      12352       average_30[0][0]                 \n",
            "__________________________________________________________________________________________________\n",
            "average_31 (Average)            (None, 728, 64)      0           lambda_101[0][0]                 \n",
            "                                                                 lambda_102[0][0]                 \n",
            "                                                                 lambda_103[0][0]                 \n",
            "                                                                 lambda_104[0][0]                 \n",
            "                                                                 lambda_105[0][0]                 \n",
            "                                                                 lambda_106[0][0]                 \n",
            "                                                                 conv1d_43[0][0]                  \n",
            "__________________________________________________________________________________________________\n",
            "global_average_pooling1d_5 (Glo (None, 64)           0           average_31[0][0]                 \n",
            "__________________________________________________________________________________________________\n",
            "ChIPseq.control.logcount (Input (None, 1)            0                                            \n",
            "__________________________________________________________________________________________________\n",
            "CUTNRUN.control.standard.logcou (None, 3)            0                                            \n",
            "__________________________________________________________________________________________________\n",
            "CUTNRUN.control.auto.logcount ( (None, 3)            0                                            \n",
            "__________________________________________________________________________________________________\n",
            "conv1d_44 (Conv1D)              (None, 654, 2)       9602        average_31[0][0]                 \n",
            "__________________________________________________________________________________________________\n",
            "ChIPseq.control.profile (InputL (None, 654, 2)       0                                            \n",
            "__________________________________________________________________________________________________\n",
            "conv1d_45 (Conv1D)              (None, 654, 2)       9602        average_31[0][0]                 \n",
            "__________________________________________________________________________________________________\n",
            "conv1d_46 (Conv1D)              (None, 654, 3)       14403       average_31[0][0]                 \n",
            "__________________________________________________________________________________________________\n",
            "CUTNRUN.control.standard.profil (None, 654, 6)       0                                            \n",
            "__________________________________________________________________________________________________\n",
            "conv1d_47 (Conv1D)              (None, 654, 3)       14403       average_31[0][0]                 \n",
            "__________________________________________________________________________________________________\n",
            "CUTNRUN.control.auto.profile (I (None, 654, 6)       0                                            \n",
            "__________________________________________________________________________________________________\n",
            "concatenate_18 (Concatenate)    (None, 65)           0           global_average_pooling1d_5[0][0] \n",
            "                                                                 ChIPseq.control.logcount[0][0]   \n",
            "__________________________________________________________________________________________________\n",
            "concatenate_19 (Concatenate)    (None, 65)           0           global_average_pooling1d_5[0][0] \n",
            "                                                                 ChIPseq.control.logcount[0][0]   \n",
            "__________________________________________________________________________________________________\n",
            "concatenate_20 (Concatenate)    (None, 67)           0           global_average_pooling1d_5[0][0] \n",
            "                                                                 CUTNRUN.control.standard.logcount\n",
            "__________________________________________________________________________________________________\n",
            "concatenate_21 (Concatenate)    (None, 67)           0           global_average_pooling1d_5[0][0] \n",
            "                                                                 CUTNRUN.control.auto.logcount[0][\n",
            "__________________________________________________________________________________________________\n",
            "concatenate_22 (Concatenate)    (None, 654, 4)       0           conv1d_44[0][0]                  \n",
            "                                                                 ChIPseq.control.profile[0][0]    \n",
            "__________________________________________________________________________________________________\n",
            "concatenate_23 (Concatenate)    (None, 654, 4)       0           conv1d_45[0][0]                  \n",
            "                                                                 ChIPseq.control.profile[0][0]    \n",
            "__________________________________________________________________________________________________\n",
            "concatenate_24 (Concatenate)    (None, 654, 9)       0           conv1d_46[0][0]                  \n",
            "                                                                 CUTNRUN.control.standard.profile[\n",
            "__________________________________________________________________________________________________\n",
            "concatenate_25 (Concatenate)    (None, 654, 9)       0           conv1d_47[0][0]                  \n",
            "                                                                 CUTNRUN.control.auto.profile[0][0\n",
            "__________________________________________________________________________________________________\n",
            "ChIPseq.POU5F1.logcount (Dense) (None, 2)            132         concatenate_18[0][0]             \n",
            "__________________________________________________________________________________________________\n",
            "ChIPseq.NANOG.logcount (Dense)  (None, 2)            132         concatenate_19[0][0]             \n",
            "__________________________________________________________________________________________________\n",
            "CUTNRUN.POU5F1.logcount (Dense) (None, 3)            204         concatenate_20[0][0]             \n",
            "__________________________________________________________________________________________________\n",
            "CUTNRUN.NANOG.logcount (Dense)  (None, 3)            204         concatenate_21[0][0]             \n",
            "__________________________________________________________________________________________________\n",
            "ChIPseq.POU5F1.profile (Conv1D) (None, 654, 2)       10          concatenate_22[0][0]             \n",
            "__________________________________________________________________________________________________\n",
            "ChIPseq.NANOG.profile (Conv1D)  (None, 654, 2)       10          concatenate_23[0][0]             \n",
            "__________________________________________________________________________________________________\n",
            "CUTNRUN.POU5F1.profile (Conv1D) (None, 654, 3)       30          concatenate_24[0][0]             \n",
            "__________________________________________________________________________________________________\n",
            "CUTNRUN.NANOG.profile (Conv1D)  (None, 654, 3)       30          concatenate_25[0][0]             \n",
            "==================================================================================================\n",
            "Total params: 128,314\n",
            "Trainable params: 128,314\n",
            "Non-trainable params: 0\n",
            "__________________________________________________________________________________________________\n",
            "None\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "stream",
          "text": [
            "W0707 14:37:04.201799 140481616963456 deprecation_wrapper.py:119] From /usr/local/lib/python3.6/dist-packages/keras/backend/tensorflow_backend.py:986: The name tf.assign_add is deprecated. Please use tf.compat.v1.assign_add instead.\n",
            "\n",
            "W0707 14:37:04.632028 140481616963456 deprecation_wrapper.py:119] From /usr/local/lib/python3.6/dist-packages/keras/backend/tensorflow_backend.py:973: The name tf.assign is deprecated. Please use tf.compat.v1.assign instead.\n",
            "\n"
          ],
          "name": "stderr"
        },
        {
          "output_type": "stream",
          "text": [
            "Epoch 1/200\n",
            "639/639 [==============================] - 403s 630ms/step - loss: 4688.1160 - ChIPseq.POU5F1.logcount_loss: 0.5575 - ChIPseq.NANOG.logcount_loss: 0.7706 - CUTNRUN.POU5F1.logcount_loss: 0.9627 - CUTNRUN.NANOG.logcount_loss: 1.1101 - ChIPseq.POU5F1.profile_loss: 108.6237 - ChIPseq.NANOG.profile_loss: 135.8883 - CUTNRUN.POU5F1.profile_loss: 11924.3883 - CUTNRUN.NANOG.profile_loss: 1263.0667 - val_loss: 608.7436 - val_ChIPseq.POU5F1.logcount_loss: 0.7734 - val_ChIPseq.NANOG.logcount_loss: 0.5639 - val_CUTNRUN.POU5F1.logcount_loss: 1.2364 - val_CUTNRUN.NANOG.logcount_loss: 1.3799 - val_ChIPseq.POU5F1.profile_loss: 103.7255 - val_ChIPseq.NANOG.profile_loss: 142.2356 - val_CUTNRUN.POU5F1.profile_loss: 380.6578 - val_CUTNRUN.NANOG.profile_loss: 483.6005\n",
            "Epoch 2/200\n",
            "639/639 [==============================] - 373s 583ms/step - loss: 2707.9039 - ChIPseq.POU5F1.logcount_loss: 0.4786 - ChIPseq.NANOG.logcount_loss: 0.6741 - CUTNRUN.POU5F1.logcount_loss: 0.8308 - CUTNRUN.NANOG.logcount_loss: 0.8404 - ChIPseq.POU5F1.profile_loss: 108.3692 - ChIPseq.NANOG.profile_loss: 135.1567 - CUTNRUN.POU5F1.profile_loss: 6577.5313 - CUTNRUN.NANOG.profile_loss: 757.3023 - val_loss: 563.9243 - val_ChIPseq.POU5F1.logcount_loss: 0.3945 - val_ChIPseq.NANOG.logcount_loss: 0.4445 - val_CUTNRUN.POU5F1.logcount_loss: 0.8089 - val_CUTNRUN.NANOG.logcount_loss: 0.7436 - val_ChIPseq.POU5F1.profile_loss: 103.4177 - val_ChIPseq.NANOG.profile_loss: 141.0429 - val_CUTNRUN.POU5F1.profile_loss: 492.7477 - val_CUTNRUN.NANOG.profile_loss: 473.5956\n",
            "Epoch 3/200\n",
            "639/639 [==============================] - 371s 581ms/step - loss: 1533.5132 - ChIPseq.POU5F1.logcount_loss: 0.4239 - ChIPseq.NANOG.logcount_loss: 0.6513 - CUTNRUN.POU5F1.logcount_loss: 0.7597 - CUTNRUN.NANOG.logcount_loss: 0.7574 - ChIPseq.POU5F1.profile_loss: 108.3268 - ChIPseq.NANOG.profile_loss: 135.0229 - CUTNRUN.POU5F1.profile_loss: 3158.1404 - CUTNRUN.NANOG.profile_loss: 688.5377 - val_loss: 479.1344 - val_ChIPseq.POU5F1.logcount_loss: 0.3689 - val_ChIPseq.NANOG.logcount_loss: 0.4441 - val_CUTNRUN.POU5F1.logcount_loss: 0.7992 - val_CUTNRUN.NANOG.logcount_loss: 0.6972 - val_ChIPseq.POU5F1.profile_loss: 103.2683 - val_ChIPseq.NANOG.profile_loss: 140.6681 - val_CUTNRUN.POU5F1.profile_loss: 262.0769 - val_CUTNRUN.NANOG.profile_loss: 463.0127\n",
            "Epoch 4/200\n",
            "639/639 [==============================] - 370s 580ms/step - loss: 1013.8544 - ChIPseq.POU5F1.logcount_loss: 0.4147 - ChIPseq.NANOG.logcount_loss: 0.6417 - CUTNRUN.POU5F1.logcount_loss: 0.7574 - CUTNRUN.NANOG.logcount_loss: 0.7564 - ChIPseq.POU5F1.profile_loss: 108.2252 - ChIPseq.NANOG.profile_loss: 134.5851 - CUTNRUN.POU5F1.profile_loss: 1655.8991 - CUTNRUN.NANOG.profile_loss: 635.9156 - val_loss: 464.7979 - val_ChIPseq.POU5F1.logcount_loss: 0.3609 - val_ChIPseq.NANOG.logcount_loss: 0.4340 - val_CUTNRUN.POU5F1.logcount_loss: 0.7987 - val_CUTNRUN.NANOG.logcount_loss: 0.6907 - val_ChIPseq.POU5F1.profile_loss: 103.2086 - val_ChIPseq.NANOG.profile_loss: 140.5400 - val_CUTNRUN.POU5F1.profile_loss: 231.4051 - val_CUTNRUN.NANOG.profile_loss: 454.7236\n",
            "Epoch 5/200\n",
            "639/639 [==============================] - 371s 580ms/step - loss: 822.0553 - ChIPseq.POU5F1.logcount_loss: 0.4077 - ChIPseq.NANOG.logcount_loss: 0.6295 - CUTNRUN.POU5F1.logcount_loss: 0.7521 - CUTNRUN.NANOG.logcount_loss: 0.7442 - ChIPseq.POU5F1.profile_loss: 108.0683 - ChIPseq.NANOG.profile_loss: 134.0045 - CUTNRUN.POU5F1.profile_loss: 1124.6828 - CUTNRUN.NANOG.profile_loss: 598.3440 - val_loss: 459.3154 - val_ChIPseq.POU5F1.logcount_loss: 0.3620 - val_ChIPseq.NANOG.logcount_loss: 0.4286 - val_CUTNRUN.POU5F1.logcount_loss: 0.7902 - val_CUTNRUN.NANOG.logcount_loss: 0.6901 - val_ChIPseq.POU5F1.profile_loss: 103.2396 - val_ChIPseq.NANOG.profile_loss: 140.4677 - val_CUTNRUN.POU5F1.profile_loss: 220.9338 - val_CUTNRUN.NANOG.profile_loss: 450.8087\n",
            "Epoch 6/200\n",
            "639/639 [==============================] - 370s 579ms/step - loss: 725.7580 - ChIPseq.POU5F1.logcount_loss: 0.4048 - ChIPseq.NANOG.logcount_loss: 0.6256 - CUTNRUN.POU5F1.logcount_loss: 0.7481 - CUTNRUN.NANOG.logcount_loss: 0.7444 - ChIPseq.POU5F1.profile_loss: 107.7360 - ChIPseq.NANOG.profile_loss: 133.4627 - CUTNRUN.POU5F1.profile_loss: 858.2474 - CUTNRUN.NANOG.profile_loss: 578.7960 - val_loss: 471.9899 - val_ChIPseq.POU5F1.logcount_loss: 0.3624 - val_ChIPseq.NANOG.logcount_loss: 0.4290 - val_CUTNRUN.POU5F1.logcount_loss: 0.8092 - val_CUTNRUN.NANOG.logcount_loss: 0.6925 - val_ChIPseq.POU5F1.profile_loss: 103.6562 - val_ChIPseq.NANOG.profile_loss: 141.8274 - val_CUTNRUN.POU5F1.profile_loss: 250.1006 - val_CUTNRUN.NANOG.profile_loss: 453.6802\n",
            "Epoch 7/200\n",
            "639/639 [==============================] - 370s 580ms/step - loss: 689.3212 - ChIPseq.POU5F1.logcount_loss: 0.4038 - ChIPseq.NANOG.logcount_loss: 0.6258 - CUTNRUN.POU5F1.logcount_loss: 0.7498 - CUTNRUN.NANOG.logcount_loss: 0.7414 - ChIPseq.POU5F1.profile_loss: 107.7056 - ChIPseq.NANOG.profile_loss: 133.1855 - CUTNRUN.POU5F1.profile_loss: 753.6989 - CUTNRUN.NANOG.profile_loss: 574.8057 - val_loss: 455.6519 - val_ChIPseq.POU5F1.logcount_loss: 0.3633 - val_ChIPseq.NANOG.logcount_loss: 0.4284 - val_CUTNRUN.POU5F1.logcount_loss: 0.7914 - val_CUTNRUN.NANOG.logcount_loss: 0.6860 - val_ChIPseq.POU5F1.profile_loss: 103.1744 - val_ChIPseq.NANOG.profile_loss: 140.3477 - val_CUTNRUN.POU5F1.profile_loss: 213.0827 - val_CUTNRUN.NANOG.profile_loss: 448.2178\n",
            "Epoch 8/200\n",
            "368/639 [================>.............] - ETA: 2:37 - loss: 642.7852 - ChIPseq.POU5F1.logcount_loss: 0.4009 - ChIPseq.NANOG.logcount_loss: 0.6167 - CUTNRUN.POU5F1.logcount_loss: 0.7411 - CUTNRUN.NANOG.logcount_loss: 0.7415 - ChIPseq.POU5F1.profile_loss: 107.6562 - ChIPseq.NANOG.profile_loss: 132.7565 - CUTNRUN.POU5F1.profile_loss: 623.0641 - CUTNRUN.NANOG.profile_loss: 569.6369"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "s_Q9qOPXbnRo",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "model.set_weights(early_stopping_callback.best_weights) #in case you terminated training midway\n",
        "model.save(\"MyProfileModel.h5\")"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "YGtlbWZZBiAQ",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "#to reload, you need to set up a CustomObjectScope\n",
        "from keras.utils import CustomObjectScope\n",
        "from keras.models import load_model\n",
        "with CustomObjectScope({'MultichannelMultinomialNLL': MultichannelMultinomialNLL}):\n",
        "  loaded_model = load_model('MyProfileModel.h5')\n",
        "  \n",
        "from google.colab import files\n",
        "files.download('MyProfileModel.h5') #download file from colab to computer"
      ],
      "execution_count": 0,
      "outputs": []
    }
  ]
}